### ISSUE TITLE
no-param-reassign: add option to ignore property assignment for whitelisted callers

### ISSUE THREAD

**Issue by dzaman** [eslint/eslint#9670]
**What rule do you want to change?**
no-param-reassign

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option - allow whitelisted (optional) object & function pairs

**Please provide some example code that this change will affect:**
There are times when it's desirable to mutate argument properties based on the caller of an iteratee. For example, a `reduce` that mutates the accumulator instead of copying can be several orders of magnitude faster.

```
array.reduce((acc, value, key) => {
  acc[key] = true;
  return acc;
}
```

It is possible to white list `acc` in `ignorePropertyModificationsFor`, but using descriptive names for accumulators is preferable.

Open to suggestions about the option name and format.

**What does the rule currently do for this code?**
Errors: Assignment to property of function parameter 'acc'.

**What will the rule do after it's changed?**
Ignore mutated arguments if the function's caller is whitelisted.

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. For what it's worth, this option was also proposed in https://github.com/eslint/eslint/issues/8581 and was not accepted.

> There are times when it's desirable to mutate argument properties based on the caller of an iteratee. For example, a `reduce` that mutates the accumulator instead of copying can be several orders of magnitude faster

I assume you're comparing to code like this, which is O(n<sup>2</sup>) time:

```js
array.reduce((acc, value, key) => ({ ...acc, [key]: true }), {})
```

However, there's a better way to do this in linear time (and similar performance to the mutation example):

```js
Object.assign({}, ...array.map((value, key) => ({ [key]: true })))
```

**Comment by zargold** [comment]
Ok but what about:
All I care about here is those 2 items the ones with names 'Gender' and 'NewItem' out of a large array... I need to extract those 2 items and sure I could loop through the large array with .find() but it is a **large** array and why should I?
I think in general with reduce you are often assigning to the object basically if you ever use an {} as the aggregator param in reduce you're probably assigning to it.
const getFilterParam = ItemsById =>
  ItemsById.reduce((items, item) => {
    /* eslint-disable no-param-reassign */
    if (item.name === 'Gender') {
      items.gender = item
    }
    if (item.name === 'NewItem') {
      items.newToSale = item
    }
    return items
    /* eslint-enable no-param-reassign */
  }, {})

**Comment by clshortfuse** [comment]
You can get around this by using `Reflect.set([target], [propertyKey], [value])` which is what I'm using for ES6 Proxies.

**Comment by zargold** [comment]
That is pretty cool and sounds safe not sure what the advantage of using this over myObj[key] = 'asdf' or myObj.definedKey = 'asdf' is there one?

- in terms of performance?
- or something else?
or is it just a way to work around eslint?

I think the real advantage of having the eslint defense against using for example the eslint rule has-own-property is that if someone accidentally uses something that isn't an object it will scream at them. 

But when you're using an object that you know is definitely an object because you just said `myObj = {}`, then the rule is not so helpful (and I'm fine doing `// eslint-disable-line no-param-reassign`) but this isn't reflecting the advantages of eslint. Eslint is great when it helps prevent you from making _potential mistakes_ and makes sure everyone keeps the same coding standards (at least IMO).

Same thing with no-param-reassign it has the intent that you not to alter arguments that are outside of this scope which might have unintended side effects. On the other hand when you just brought in an empty object to perform mutations on it you don't care whether anything will be assigned to because again you just created it and assigned it an empty value. It is not a real param that you are modifying in your function it is a placeholder that you put in (when you use reduce).

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
